proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,305 @@
+one or more _upstream clusters_.    An _upstream cluster_ represents the physical kafka cluster.  There is a one to one correspondance between
+a _upstream cluster_ and a physical cluster in the world.
+
+There will be a mechansim to route flexibly between downstreams and upstreams.  This will allow useful topologies to built to serve different use-cases.
+
+- *one to one* - the simple exposure of a physical cluster 
+- *many to one* - kroxy presents many downstream clusters which route to a single upstream - this would support a multitenant use-cases 
+
+(There's a possiblity of *one-to-many* which a single downstream routes to several upstreams as if they were one, which might be useful for presenting several
+kafka clusters as if they were one, however, supporting transactions across two clusters would be difficult.  We consider won't this use-case futrther).
+
+## Current situation
+
+Currently Kroxyilicious binds a single port to the downstream and connects to a single upstream broker.  There is no support for clusters comprised of more
+then one broker.
+
+## Motivation
+
+1. Ability for Kroxylicious to expose multiple physical kafka clusters.
+1. Ability for Kroxylicious to support clusters formed of more than one broker.
+1. Ability for Kroxylicious to provide a building block to support multi-tenancy.
+
+## Proposal
+
+The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.
+
+### Key Objects
+
+
+![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)
+
+
+
+
+
+#### Endpoints
+
+Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  
+
+* For the non-TLS case, multiple endpoints are required to support >1 brokers. 
+* For TLS case, SNI gives a mechanism to route the traffic to a >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.
+
+`ProxyConfig` will be replaced by an list of _Endpoint_ objects.  
+
+```
+- name: endpoint name
+  address: bind address/port
+  type: PLAIN|TLS
+```
+
+#### Downstream Clusters
+
+Conceptually the _downstream cluster_ represents the cluster that the kafka client connects to.  The downstream clusters list specifies
+all the kafka clustsers are being presented by Kroxylicious. 
+
+A _downstream clustser_ references exactly one _upstream cluster_. 
+
+A _downstream cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the upstream cluster.
+
+A _downstream cluster_ enumerates the downstream brokers that comprise the downstream cluster.
```

Comment:
Could we add something here explicitly stating the relationship between Downstream Cluster and Endpoint?

Something like `A downstream cluster is associated with a single Endpoint. Multiple downstream clusters may refer to the same Endpoint`

Same with TLS like `A downstream cluster can provide TLS configuration which is used for XYZ`